src/tunnel.cc

```diff
@@ -430,16 +433,16 @@ TunnelStateData::retryOrBail(const char *context)
     // sendNewError() and sendSavedErrorOr(), used in "error detected" cases.
     const auto error = savedError ? savedError : new ErrorState(ERR_CANNOT_FORWARD,
             Http::scInternalServerError, request.getRaw(), al);
-    if (Comm::IsConnOpen(client.conn) && clientExpectsConnectResponse()) {
-        sendError(error, context);
-        return;
-    }
+    const auto canSendError = Comm::IsConnOpen(client.conn) && !client.dirty &&
```

Comment:
I am trying to understand whether the new Connection::dirty is really required: can we use the existing 'retriable' flag here? It becomes 'false' in notePeerReadyToShovel(), whereas 'durty' I guess becomes true a bit later, somewhere in tunnelStartShoveling().

Reply:
I rejected the idea of reusing `retriable` (i.e. `dirty == !retriable`) because the two states have different (and largely orthogonal) semantics. Regardless of what the code does (not do) now, it may be possible, in theory, to retry the connection to the _next_ hop after the connection to the _previous_ hop becomes dirty. In other words, I can image some future code having true `retriable` (yes, can try another peer) and true `dirty` (no, can no longer respond with an error).

The above logic does not necessarily imply that `retriable` is falsified too early today. Squid may not be able and/or may not be _willing_ to retry Squid-to-peer failures after notePeerReadyToShovel(). Imagine, for example, a write timeout for the first chunk of data going from Squid to the peer. We could retry after that error (at least in some cases) because no client data was sent to the next hop, but there is currently no code to support such retries.

Reply:
Ok.